### PR TITLE
Feature/manage user roles redux

### PR DESCRIPTION
**Pull request by DavidOberholzer**
**Background:** The Manage User Roles page was implemented prior to understanding how to hook into the Admin on Rest Redux Store. Also the TableField component looked different on the Manage User Roles page and was not able to be limited in the case of an excessive amount of results in the table. 

**What was done:** The store was updated to include the `manageUserRoles` state, which governs the Manage User Roles page. This update includes actionTypes, actions and a reducer to cover all of the existing functionality. All the Manage User Role components were connected to the store and rely on the current Redux state for there rendering now. This reduced a lot of visual complexity of the code and also a seemly better performance :fireworks: The TableField was also updated with a `limit` props to limit the amount shown in the table and the outstanding amount shown below the table as information. A props `useCard` was also given to the TableField in order to use a material-ui `Card` as the outer component rather than a div as required in the Manage User Roles page. Some Styles were updated and when a Card is used for the table a `CardTitle` is used for the heading rather than just a `label`. Attached is a screen shot of the table in the Manage User Roles page and  then in a Domain View page with `useCard` set to false (default behaviour).

screenshot from 2018-07-03 13-32-45
screenshot from 2018-07-03 13-33-16

### CODE REVIEW

File: admin/src/pages/ManageUserRoles/AssignRoleCard.js [hunk 1]
```diff
@@ -7,64 +8,179 @@ import CardTitle from 'material-ui/Card/CardTitle';
...
+        let successCount = amountSelectedToAssign;
+        if (successCount) {
+            this.setState({ assigning: true });
+            const [placeType, placeID] = assignmentLocation.split(':');
+            const place = placeType === 'd' ? 'domain' : 'site';
+            let count = 0;
+            Object.values(rolesToAssign).map((role, index) => {
+                if (role.checked) {
+                    count += 1;
+                    restClient(CREATE, `user${place}roles`, {
```
Comment (chrislombaard) [inline admin/src/pages/ManageUserRoles/AssignRoleCard.js]: Is there no way you can wrap this in an action?

Comment (chrislombaard) [reply admin/src/pages/ManageUserRoles/AssignRoleCard.js]: By that I mean move it out to its `actions.js` file

File: admin/src/pages/ManageUserRoles/index.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+    componentDidMount() {
+        if (!this.props.manageUserRoles.roleMapping) {
+            this.setupManageUserRolesStore();
+        }
+    }
+
+    setupManageUserRolesStore() {
+        /*
+			This method will initialize the store of the
```
Comment (CSergienko) [inline admin/src/pages/ManageUserRoles/index.js]: 🐒 Indentation looks wonky.

File: admin/src/pages/ManageUserRoles/AssignRoleCard.js [hunk 3]
```diff
@@ -7,64 +8,179 @@ import CardTitle from 'material-ui/Card/CardTitle';
...
+                        value={assignmentLocation}
+                        onChange={this.handleChange}
+                        onlyDomains={false}
+                        useReduxFormField={false}
+                    />
+                    {assignmentLocation && (
+                        <div>
+                            <CardHeader subtitle="Please choose the roles to add:" />
+                            <CardText>
+                                {notEmptyObject(rolesToAssign)
```
Comment (chrislombaard) [inline admin/src/pages/ManageUserRoles/AssignRoleCard.js]: Why not just `rolesToAssign ? `

File: admin/src/pages/ManageUserRoles/UserCard.js [hunk 4]
```diff
@@ -9,83 +10,165 @@ import Divider from 'material-ui/Divider';
...
+        this.handleClose = this.handleClose.bind(this);
+        this.handleDelete = this.handleDelete.bind(this);
+        this.handleAPIError = this.handleAPIError.bind(this);
+    }
+
+    handleCheck(key) {
+        this.props.checkRoleForDelete(key);
+    }
+
+    triggerDeleteDialog() {
```
Comment (chrislombaard) [inline admin/src/pages/ManageUserRoles/UserCard.js]: Try and stick to one convention for referring to `opening` or `triggering` or `handling` 
a dialog.

File: admin/src/pages/ManageUserRoles/UserCard.js [hunk 5]
```diff
@@ -9,83 +10,165 @@ import Divider from 'material-ui/Divider';
...
+                                          userRole.role.label
+                                      }' from '${userRole[resource].name}'.`
+                                    : `Something went wrong. Cannot delete role '${
+                                          userRole.role.label
+                                      }' from '${userRole[resource].name}' for user`;
+                            errorNotificationAnt(description);
+                            this.handleAPIError(error);
+                        });
+                }
+                return null;
```
Comment (chrislombaard) [inline admin/src/pages/ManageUserRoles/UserCard.js]: You don't need to return `null` here

File: admin/src/pages/ManageUserRoles/UserCard.js [hunk 6]
```diff
@@ -9,83 +10,165 @@ import Divider from 'material-ui/Divider';
...
+        }
+    }
+
+    handleAPIError(error) {
+        if (error.message === 'Token expired') {
+            localStorage.removeItem('id_token');
+            localStorage.removeItem('permissions');
+            this.props.invalidToken();
+        }
+        throw new Error(error);
```
Comment (chrislombaard) [inline admin/src/pages/ManageUserRoles/UserCard.js]: Throwing an error here is redundant.

Comment (chrislombaard) [reply admin/src/pages/ManageUserRoles/UserCard.js]: You've already caught the `error `in question

File: admin/src/pages/ManageUserRoles/AssignRoleCard.js [hunk 7]
```diff
@@ -7,64 +8,179 @@ import CardTitle from 'material-ui/Card/CardTitle';
...
+                        }
+                    })
+                        .then(response => {
+                            successCount -= 1;
+                            const placeObject =
+                                place === 'domain'
+                                    ? managerDomains[`d:${placeID}`]
+                                    : managerSites[`s:${placeID}`];
+                            this.props.assignRole(`${placeType}:${placeID}:${role.id}`, {
+                                [place]: placeObject,
```
[the remark this answers is not in the document]
Comment (DavidOberholzer) [reply admin/src/pages/ManageUserRoles/AssignRoleCard.js]: its when you use a variable value as the key name

File: admin/src/pages/ManageUserRoles/UserSearch.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        });
+    }
+
+    handleSearch(event) {
+        const value = event.target.value;
+        this.setState({
+            value
+        });
+        if (value.length > 2) {
+            restClient(GET_LIST, 'users', {
```
Comment (chrislombaard) [inline admin/src/pages/ManageUserRoles/UserSearch.js]: consider refactoring these kinds of calls out into their own functions.

Comment (chrislombaard) [reply admin/src/pages/ManageUserRoles/UserSearch.js]: Will make these easier to read

File: admin/src/pages/ManageUserRoles/index.js (hunk 2 above)
Comment (chrislombaard) [inline admin/src/pages/ManageUserRoles/index.js]: Formatting?

File: admin/src/pages/ManageUserRoles/index.js [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        this.handleAPIError = this.handleAPIError.bind(this);
+    }
+
+    componentDidMount() {
+        if (!this.props.manageUserRoles.roleMapping) {
+            this.setupManageUserRolesStore();
+        }
+    }
+
+    setupManageUserRolesStore() {
```
Comment (chrislombaard) [inline admin/src/pages/ManageUserRoles/index.js]: This function definitely needs to be refactored into smaller pieces for each step

File: admin/src/fields/TableField.js [hunk 10]
```diff
@@ -10,9 +14,88 @@ import {
...
+                <CardText style={styles.customTableLabel}>
+                    <label>
+                        <span>{label}</span>
+                    </label>
+                </CardText>
+            )}
+            <Table selectable={selectable} onRowSelection={onRowSelection}>
+                <TableHeader displaySelectAll={false} adjustForCheckbox={selectable}>
+                    <TableRow>
+                        {Object.keys(data[0]).map(header => (
```
[the remark this answers is not in the document]
Comment (DavidOberholzer) [reply admin/src/fields/TableField.js]: The component will not render the table without any entries in `data`

File: admin/src/pages/ManageUserRoles/UserCard.js (hunk 5 above)
Comment (DavidOberholzer) [reply admin/src/pages/ManageUserRoles/UserCard.js]: the compilation complains that nothing is returned in the mapping. It clears the warning

File: admin/src/pages/ManageUserRoles/UserCard.js [hunk 11]
```diff
@@ -9,83 +10,165 @@ import Divider from 'material-ui/Divider';
...
+        this.setState({ open: false });
+    }
+
+    handleDelete() {
+        const { selectedUser, userResults, userRoles } = this.props.manageUserRoles;
+        const user = userResults[selectedUser];
+        if (user) {
+            Object.entries(userRoles).map(([key, userRole]) => {
+                if (userRole.checked) {
+                    const resource = key.startsWith('d') ? 'domain' : 'site';
```
[the remark this answers is not in the document]
Comment (DavidOberholzer) [reply admin/src/pages/ManageUserRoles/UserCard.js]: At the moment, yes

File: admin/src/pages/ManageUserRoles/index.js (hunk 2 above)
Comment (DavidOberholzer) [reply admin/src/pages/ManageUserRoles/index.js]: Hmm will fix

File: admin/src/actions/manageUserRoles.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
+export const MANAGE_USER_ROLES_SET_ROLE_MAPPING = 'MANAGE_USER_ROLES_SET_ROLE_MAPPING';
```
[the remark this answers is not in the document]
Comment (DavidOberholzer) [reply admin/src/actions/manageUserRoles.js]: Changed

File: admin/src/fields/TableField.js [hunk 13]
```diff
@@ -10,9 +14,88 @@ import {
...
+                <CardTitle title={label} />
+            ) : (
+                <CardText style={styles.customTableLabel}>
+                    <label>
+                        <span>{label}</span>
+                    </label>
+                </CardText>
+            )}
+            <Table selectable={selectable} onRowSelection={onRowSelection}>
+                <TableHeader displaySelectAll={false} adjustForCheckbox={selectable}>
```
[the remark this answers is not in the document]
Comment (DavidOberholzer) [reply admin/src/fields/TableField.js]: it's default is true

File: admin/src/actions/manageUserRoles.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (jwfwessels) [reply admin/src/actions/manageUserRoles.js]: 🐒 
agreed, but nitpick

File: admin/src/fields/TableField.js [hunk 14]
```diff
@@ -10,9 +14,87 @@ import {
...
+        linkField,
+        linkedResource,
+        onRowSelection,
+        selectable,
+        selected,
+        useCard
+    } = props;
+    return (
+        <React.Fragment>
+            {useCard ? (
```
Comment (jwfwessels) [inline admin/src/fields/TableField.js]: whats this distinction for?

File: admin/src/pages/ManageUserRoles/AssignRoleCard.js (hunk 3 above)
Comment (DavidOberholzer) [reply admin/src/pages/ManageUserRoles/AssignRoleCard.js]: Empty object is truthy

File: admin/src/fields/TableField.js [hunk 15]
```diff
@@ -10,9 +14,87 @@ import {
     TableRow,
     TableRowColumn
 } from 'material-ui/Table';
+
 import restClient, { OPERATIONAL } from '../swaggerRestServer';
 import { styles } from '../Theme';
 
+const GenerateTable = ({ data, props }) => {
```
Comment (jwfwessels) [inline admin/src/fields/TableField.js]: why did you feel the need to separate data from the other props? and then destructure props separately on the next line?

File: admin/src/fields/TableField.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (jwfwessels) [reply admin/src/fields/TableField.js]: @DavidOberholzer that check does not happen around the header row though 🤔 ?

File: admin/src/fields/TableField.js [hunk 16]
```diff
@@ -10,9 +14,88 @@ import {
...
+                <TableHeader displaySelectAll={false} adjustForCheckbox={selectable}>
+                    <TableRow>
+                        {Object.keys(data[0]).map(header => (
+                            <TableHeaderColumn key={header} style={styles.customTableHeader}>
+                                {header.toUpperCase()}
+                            </TableHeaderColumn>
+                        ))}
+                    </TableRow>
+                </TableHeader>
+                <TableBody displayRowCheckbox={selectable} showRowHover={true}>
```
[the remark this answers is not in the document]
Comment (jwfwessels) [reply admin/src/fields/TableField.js]: @chrislombaard 
I do prefer the explicitness of declaring it though. else it sometimes becomes hard to reason about weather its a incoming prop or a locally defined one 😄

File: admin/src/fields/TableField.js [hunk 17]
```diff
@@ -10,9 +14,87 @@ import {
...
+                    {data.map((entry, index) => {
+                        const isSelected = onRowSelection ? selected === index : null;
+                        return (
+                            <TableRow key={index} selected={isSelected}>
```
Comment (jwfwessels) [inline admin/src/fields/TableField.js]: if you find that your component is becoming hard to manage then beard is correct in saying that you can easily break the TableRow out and declare it above and just call it here.

File: admin/src/fields/TableField.js [hunk 18]
```diff
@@ -10,9 +14,88 @@ import {
...
+                        {Object.keys(data[0]).map(header => (
+                            <TableHeaderColumn key={header} style={styles.customTableHeader}>
+                                {header.toUpperCase()}
+                            </TableHeaderColumn>
+                        ))}
+                    </TableRow>
+                </TableHeader>
+                <TableBody displayRowCheckbox={selectable} showRowHover={true}>
+                    {data.map((entry, index) => {
+                        const isSelected = onRowSelection ? selected === index : null;
```
[the remark this answers is not in the document]
Comment (jwfwessels) [reply admin/src/fields/TableField.js]: I think you can actaully just check if `selected === index`? and do it right in the prop def below?

File: admin/src/fields/TableField.js [hunk 19]
```diff
@@ -10,9 +14,88 @@ import {
...
+                        return (
+                            <TableRow key={index} selected={isSelected}>
+                                {Object.keys(entry).map((header, index) => (
+                                    <TableRowColumn key={index}>
+                                        {entry[header] instanceof Array ? (
+                                            <div style={styles.wrapper}>
+                                                {entry[header].map(
+                                                    (str, index) =>
+                                                        str ? (
```
[the remark this answers is not in the document]
Comment (jwfwessels) [reply admin/src/fields/TableField.js]: why would the header map contain falsy values?

File: admin/src/fields/TableField.js [hunk 20]
```diff
@@ -10,9 +14,87 @@ import {
...
+                            </TableRow>
+                        );
+                    })}
+                </TableBody>
+            </Table>
+            {limit &&
+                data.length > limit && (
+                    <CardHeader
+                        style={styles.customTableBottom}
+                        subtitle={`${data.length - limit} more found...`}
```
Comment (jwfwessels) [inline admin/src/fields/TableField.js]: hm... is there a reason for introducing this limit? why not just paginate the results?

File: admin/src/fields/TableField.js [hunk 21]
```diff
@@ -26,19 +108,16 @@ class TableField extends Component {
     }
```
Comment (jwfwessels) [inline admin/src/fields/TableField.js]: 🐒 
clean this up please ;) ^ the console log in your `componentWillMount`

File: admin/src/fields/TableField.js [hunk 22]
```diff
@@ -48,83 +128,18 @@ class TableField extends Component {
...
-            </div>
+        const { data } = this.state;
+        console.log(this.props);
+        return data.length ? (
+            this.props.useCard ? (
+                <Card style={styles.customTableDiv}>
+                    <GenerateTable data={data} props={this.props} />
+                </Card>
+            ) : (
+                <div style={styles.customTableDiv}>
```
[the remark this answers is not in the document]
Comment (jwfwessels) [reply admin/src/fields/TableField.js]: curious as to what you are trying to do here?

File: admin/src/pages/ManageUserRoles/AssignRoleCard.js [hunk 23]
```diff
@@ -7,64 +8,179 @@ import CardTitle from 'material-ui/Card/CardTitle';
...
-        </Card>
-    );
-};
+const mapStateToProps = state => ({
+    manageUserRoles: state.manageUserRoles
+});
 
-export default AssignRoleCard;
+const mapDispatchToProps = dispatch => ({
+    invalidToken: () => dispatch(invalidToken()),
```
Comment (jwfwessels) [inline admin/src/pages/ManageUserRoles/AssignRoleCard.js]: all of these can be written in shorthand 😄

File: admin/src/pages/ManageUserRoles/AssignRoleCard.js [hunk 24]
```diff
@@ -7,64 +8,179 @@ import CardTitle from 'material-ui/Card/CardTitle';
...
+                            this.props.assignRole(`${placeType}:${placeID}:${role.id}`, {
+                                [place]: placeObject,
+                                role: roleMapping[role.id],
+                                checked: false
+                            });
+                            successNotificationAnt(
+                                `Role '${role.label}' assigned on ${place} '${placeObject.name}'`,
+                                3
+                            );
+                            if (!successCount) {
```
Comment (jwfwessels) [inline admin/src/pages/ManageUserRoles/AssignRoleCard.js]: 🐒 
I feel this section with the notifications is a bit finicky 🤔

File: admin/src/fields/TableField.js [hunk 25]
```diff
@@ -10,9 +14,88 @@ import {
...
+                        ))}
+                    </TableRow>
+                </TableHeader>
+                <TableBody displayRowCheckbox={selectable} showRowHover={true}>
+                    {data.map((entry, index) => {
+                        const isSelected = onRowSelection ? selected === index : null;
+                        return (
+                            <TableRow key={index} selected={isSelected}>
+                                {Object.keys(entry).map((header, index) => (
+                                    <TableRowColumn key={index}>
```
[the remark this answers is not in the document]
Comment (DavidOberholzer) [reply admin/src/fields/TableField.js]: I tried to no but it errors stating that the children require to be that of material-ui. Making parent components to house smaller portions causes it to not render and throw a console error from material-ui

File: admin/src/fields/TableField.js [hunk 26]
```diff
@@ -10,9 +14,88 @@ import {
...
+            ) : (
+                <CardText style={styles.customTableLabel}>
+                    <label>
+                        <span>{label}</span>
+                    </label>
+                </CardText>
+            )}
+            <Table selectable={selectable} onRowSelection={onRowSelection}>
+                <TableHeader displaySelectAll={false} adjustForCheckbox={selectable}>
+                    <TableRow>
```
[the remark this answers is not in the document]
Comment (DavidOberholzer) [reply admin/src/fields/TableField.js]: same here

File: admin/src/pages/ManageUserRoles/UserSearch.js (hunk 8 above)
Comment (DavidOberholzer) [reply admin/src/pages/ManageUserRoles/UserSearch.js]: Will note that and then do in a separate PR in all of my components for the project, adding as TODO

File: admin/src/actions/manageUserRoles.js [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+    MANAGE_USER_ROLES_SELECT_USER,
+    MANAGE_USER_ROLES_CHECK_ROLE_FOR_DELETE,
+    MANAGE_USER_ROLES_DELETE_ROLE,
+    MANAGE_USER_ROLES_SET_ASSIGNMENT_LOCATION,
+    MANAGE_USER_ROLES_CHECK_ROLE_FOR_ASSIGN,
+    MANAGE_USER_ROLES_ASSIGN_ROLE,
+    MANAGE_USER_ROLES_ALL_ASSIGNED,
+    MANAGE_USER_ROLES_RESET,
+    MANAGE_USER_ROLES_INVALID_TOKEN
+} from '../actionTypes';
```
Comment (chrislombaard) [inline admin/src/actions/manageUserRoles.js]: Noice

File: admin/src/pages/ManageUserRoles/UserCard.js [hunk 28]
```diff
@@ -9,83 +10,165 @@ import Divider from 'material-ui/Divider';
...
-    handleCheck: PropTypes.func.isRequired,
-    triggerDeleteDialog: PropTypes.func.isRequired
-};
+const mapStateToProps = state => ({
+    manageUserRoles: state.manageUserRoles
+});
 
-export default UserCard;
+const mapDispatchToProps = dispatch => ({
+    checkRoleForDelete: key => dispatch(checkRoleForDelete(key)),
```
Comment (jwfwessels) [inline admin/src/pages/ManageUserRoles/UserCard.js]: 🐒 
can be written in shorthand

File: admin/src/actionTypes.js [hunk 29]
```diff
@@ -0,0 +1,19 @@
+export const CONTEXT_DOMAINS_AND_SITES_ADD = 'CONTEXT_DOMAINS_AND_SITES_ADD';
```
Comment (chrislombaard) [inline admin/src/actionTypes.js]: You don't need to explicitly export each one, you can just do 
```export const ACTION_1 = 'ACTION_1',
                            ACTION_2 = 'ACTION_2';
```

File: admin/src/pages/ManageUserRoles/UserCard.js (hunk 5 above)
Comment (jwfwessels) [reply admin/src/pages/ManageUserRoles/UserCard.js]: you can just say return @DavidOberholzer  👍

File: admin/src/inputs/DomainTreeInput.js [hunk 30]
```diff
@@ -53,12 +53,12 @@ class DomainTreeInput extends Component {
                             this.setState({ treeData: createTreeData(places, 'domain_id', 's') });
                         })
                         .catch(error => {
-                            throw new Error(error);
+                            console.error(error);
```
Comment (chrislombaard) [inline admin/src/inputs/DomainTreeInput.js]: We should still not have console outputs. Is this used to debug?

File: admin/src/fields/TableField.js (hunk 20 above)
Comment (DavidOberholzer) [reply admin/src/fields/TableField.js]: It was an afterthought to add this, it isnt ticketed anywhere. Maybe I can improve on it in a different PR

File: admin/src/inputs/DomainTreeInput.js (hunk 30 above)
Comment (DavidOberholzer) [reply admin/src/inputs/DomainTreeInput.js]: Yea just and easier way to know it was a call that went wrong

File: admin/src/actionTypes.js (hunk 29 above)
Comment (jwfwessels) [reply admin/src/actionTypes.js]: @chrislombaard ^ thats a preference thing tbh. I prefer doing each line. its much of a muchness

File: admin/src/pages/ManageUserRoles/UserCard.js (hunk 5 above)
Comment (DavidOberholzer) [reply admin/src/pages/ManageUserRoles/UserCard.js]: Still complains, must return a value

**Pull request merged**